Injection (Component, Token) fails during building (3.0.0-rc.3, also ^2.0.0) in Angular 6.0

## Type of Issue

```
[x] Bug Report
[ ] Feature Request
```

## Description
While upgrading my UI library to Angular 6 I encountered a problem with Injection during the build with ng-packagr (Error Message in appended and inside the minimal example)

Here is the most minimal example I was able to create.
https://github.com/georgiee/ng-packagr-inject
(I tried to use stackblitz so you can quickly take a look- but it's just not made for CLI stuff)

The weird thing is: Injection of the Token and the Component Reference happens in two different places. But once I remove either of them the problem goes away.

I can imagine that this is a ngc problem as it fails during the step `Compiling TypeScript sources through ngc` inside the file `ng-packagr/lib/ngc/compile-source-files.js` but I'm unsure so that I rather open this issue.

I don't know if this is related:
https://github.com/angular/angular/issues/23734
A quick `@Injectable({providedIn: 'root'})` did not help in my example.

#### How To Reproduce
Just build the library in the example project. I also wanted to try a raw ngc example but I have no real clue where to begin. I want to drop this issue here instead in case you see what's going on.

#### Expected Behaviour
The build should be fine in this easy scenario.

#### Version Information

```
ng-packagr:            3.0.0-rc.3
@angular/compiler:     6.0.0
rollup:                0.58.2
tsickle:               0.27.5
typescript:            2.7.2
```



Here the error details:


> Warning: Can't resolve all parameters for FooClass in ng-packagr-bug-inject/projects/library/foo/public-api.ts: (?). This will become an error in Angular v6.x

```
BUILD ERROR
projects/dist/library/bar/public-api.d.ts(1,31): error TS2307: Cannot find module '@my/library/some-component'.
: Can't resolve all parameters for BarDirective inng-packagr-bug-inject/projects/dist/library/bar/my-library-bar.d.ts: (?).

Error: projects/dist/library/bar/public-api.d.ts(1,31): error TS2307: Cannot find module '@my/library/some-component'.
: Can't resolve all parameters for BarDirective in ng-packagr-bug-inject/projects/dist/library/bar/my-library-bar.d.ts: (?).

    at Object.<anonymous> (ng-packagr-bug-inject/node_modules/ng-packagr/lib/ngc/compile-source-files.js:62:68)
    at Generator.next (<anonymous>)
    atng-packagr-bug-inject/node_modules/ng-packagr/lib/ngc/compile-source-files.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (ng-packagr-bug-inject/node_modules/ng-packagr/lib/ngc/compile-source-files.js:3:12)
    at Object.compileSourceFiles (ng-packagr-bug-inject/node_modules/ng-packagr/lib/ngc/compile-source-files.js:21:12)
    at Object.<anonymous> (ng-packagr-bug-inject/node_modules/ng-packagr/lib/ng-v5/entry-point/ts/compile-ngc.transform.js:42:34)
    at Generator.next (<anonymous>)
    atng-packagr-bug-inject/node_modules/ng-packagr/lib/ng-v5/entry-point/ts/compile-ngc.transform.js:7:71
    at new Promise (<anonymous>)
```

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -100,6 +100,13 @@
     {
       "type": "node",
       "request": "launch",
+      "name": "Sample: issue-852",
+      "program": "${workspaceRoot}/integration/samples.dev.js",
+      "args": ["issue-852"]
+    },
+    {
+      "type": "node",
+      "request": "launch",
       "name": "Sample: externals",
       "program": "${workspaceRoot}/integration/samples.dev.js",
       "args": ["externals"]

--- a/integration/samples/issue-852/bar/package.json
+++ b/integration/samples/issue-852/bar/package.json
@@ -1,0 +1,7 @@
+{
+  "ngPackage": {
+    "lib": {
+      "entryFile": "public-api.ts"
+    }
+  }
+}

--- a/integration/samples/issue-852/bar/public-api.ts
+++ b/integration/samples/issue-852/bar/public-api.ts
@@ -1,0 +1,17 @@
+import { SomeComponent } from '@my/issue-852/some-component';
+import { Directive, InjectionToken, NgModule, Optional } from '@angular/core';
+
+export const BAR_TOKEN = new InjectionToken<string>('bar-token');
+
+@Directive({
+  selector: 'input[Bar]'
+})
+export class BarDirective {
+  constructor(@Optional() private _formField: SomeComponent) {}
+}
+
+@NgModule({
+  declarations: [BarDirective],
+  exports: [BarDirective]
+})
+export class BarModule {}

--- a/integration/samples/issue-852/foo/package.json
+++ b/integration/samples/issue-852/foo/package.json
@@ -1,0 +1,7 @@
+{
+  "ngPackage": {
+    "lib": {
+      "entryFile": "public-api.ts"
+    }
+  }
+}

--- a/integration/samples/issue-852/foo/public-api.ts
+++ b/integration/samples/issue-852/foo/public-api.ts
@@ -1,0 +1,12 @@
+import { Platform } from '@angular/cdk/platform';
+import { Inject, Injectable, Optional } from '@angular/core';
+import { BAR_TOKEN } from '@my/issue-852/bar';
+
+@Injectable()
+export class FooClass {
+  constructor(
+    @Optional()
+    @Inject(BAR_TOKEN)
+    matDateLocale: string
+  ) {}
+}

--- a/integration/samples/issue-852/package.json
+++ b/integration/samples/issue-852/package.json
@@ -1,0 +1,13 @@
+{
+  "name": "@my/issue-852",
+  "private": true,
+  "version": "0.0.0",
+  "scripts": {
+    "test": "../../../node_modules/.bin/cross-env TS_NODE_PROJECT=../../../integration/tsconfig.specs.json ../../../node_modules/.bin/mocha --compilers ts:ts-node/register specs/**/*.ts"
+  },
+  "ngPackage": {
+    "lib": {
+      "entryFile": "public-api.ts"
+    }
+  }
+}

--- a/integration/samples/issue-852/public-api.ts
+++ b/integration/samples/issue-852/public-api.ts
@@ -1,0 +1,1 @@
+export const SOME_VALUE = 'hello';

--- a/integration/samples/issue-852/some-component/package.json
+++ b/integration/samples/issue-852/some-component/package.json
@@ -1,0 +1,7 @@
+{
+  "ngPackage": {
+    "lib": {
+      "entryFile": "public-api.ts"
+    }
+  }
+}

--- a/integration/samples/issue-852/some-component/public-api.ts
+++ b/integration/samples/issue-852/some-component/public-api.ts
@@ -1,0 +1,15 @@
+import { Component, NgModule } from '@angular/core';
+
+@Component({
+  selector: 'some-component',
+  template: ''
+})
+export class SomeComponent {
+  constructor() {}
+}
+
+@NgModule({
+  declarations: [SomeComponent],
+  exports: [SomeComponent]
+})
+export class SomeComponentModule {}

--- a/integration/samples/issue-852/spec/metadata.ts
+++ b/integration/samples/issue-852/spec/metadata.ts
@@ -1,0 +1,28 @@
+import { expect } from 'chai';
+import * as fs from 'fs';
+import * as path from 'path';
+
+describe(`issue-852`, () => {
+  describe(`issue-852.metadata.json`, () => {
+    let METADATA;
+    before(() => {
+      METADATA = require('../dist/my-issue-852.metadata.json');
+    });
+
+    it(`should exist`, () => {
+      expect(METADATA).to.be.ok;
+    });
+
+    it(`should be version 4`, () => {
+      expect(METADATA.version).to.equal(4);
+    });
+
+    it(`should be "__symbolic": "module"`, () => {
+      expect(METADATA['__symbolic']).to.equal('module');
+    });
+
+    it(`should "importAs": "@my/issue-852"`, () => {
+      expect(METADATA['importAs']).to.equal('@my/issue-852');
+    });
+  });
+});

--- a/src/lib/ng-v5/entry-point/ts/compile-ngc.transform.ts
+++ b/src/lib/ng-v5/entry-point/ts/compile-ngc.transform.ts
@@ -18,26 +18,6 @@ export const compileNgcTransform: Transform = transformFromPromise(async graph =
   const tsSources = entryPoint.find(isTypeScriptSources) as TypeScriptSourceNode;
   const tsConfig: TsConfig = entryPoint.data.tsConfig;
 
-  // Add paths mappings for dependencies
-  const entryPointDeps = entryPoint.filter(isEntryPoint) as EntryPointNode[];
-  if (entryPointDeps.length > 0) {
-    if (!tsConfig.options.paths) {
-      tsConfig.options.paths = {};
-    }
-
-    for (let dep of entryPointDeps) {
-      const { entryPoint, destinationFiles } = dep.data;
-      const depModuleId = entryPoint.moduleId;
-      const mappedPath = [destinationFiles.declarations];
-
-      if (!tsConfig.options.paths[depModuleId]) {
-        tsConfig.options.paths[depModuleId] = mappedPath;
-      } else {
-        tsConfig.options.paths[depModuleId].concat(mappedPath);
-      }
-    }
-  }
-
   // Compile TypeScript sources
   const { esm2015, esm5, declarations } = entryPoint.data.destinationFiles;
   const previousTransform = tsSources.data;

--- a/src/lib/ng-v5/init/init-tsconfig.transform.ts
+++ b/src/lib/ng-v5/init/init-tsconfig.transform.ts
@@ -2,18 +2,12 @@
 import { ParsedConfiguration } from '@angular/compiler-cli/src/perform_compile';
 import { Transform, transformFromPromise } from '../../brocc/transform';
 import { TsConfig, initializeTsConfig } from '../../ts/tsconfig';
-import * as log from '../../util/log';
-import { isEntryPoint } from '../nodes';
+import { isEntryPoint, EntryPointNode } from '../nodes';
 
 export const initTsConfigTransformFactory = (defaultTsConfig: TsConfig): Transform =>
   transformFromPromise(async graph => {
     // Initialize tsconfig for each entry point
-    const entryPoints = graph.filter(isEntryPoint);
-    for (let entryPoint of entryPoints) {
-      log.debug(`Initializing tsconfig for ${entryPoint.data.entryPoint.moduleId}`);
-      const tsConfig = initializeTsConfig(defaultTsConfig, entryPoint.data.entryPoint);
-      entryPoint.data.tsConfig = tsConfig;
-    }
-
+    const entryPoints = graph.filter(isEntryPoint) as EntryPointNode[];
+    initializeTsConfig(defaultTsConfig, entryPoints);
     return graph;
   });

--- a/src/lib/ts/tsconfig.ts
+++ b/src/lib/ts/tsconfig.ts
@@ -4,6 +4,8 @@ import { ParsedConfiguration } from '@angular/compiler-cli/src/perform_compile';
 import * as path from 'path';
 import * as ts from 'typescript';
 import { NgEntryPoint } from '../ng-package-format/entry-point';
+import { EntryPointNode } from '../ng-v5/nodes';
+import * as log from '../util/log';
 
 /**
  * TypeScript configuration used internally (marker typer).
@@ -40,59 +42,83 @@ export function createDefaultTsConfig(values?: TsConfig | string): TsConfig {
  * Initializes TypeScript Compiler options and Angular Compiler options by overriding the
  * default config with entry point-specific values.
  */
-export const initializeTsConfig = (defaultTsConfig: TsConfig, entryPoint: NgEntryPoint): TsConfig => {
-  const basePath = path.dirname(entryPoint.entryFilePath);
+export const initializeTsConfig = (defaultTsConfig: TsConfig, entryPoints: EntryPointNode[]) => {
+  entryPoints.forEach(currentEntryPoint => {
+    const { entryPoint } = currentEntryPoint.data;
+    log.debug(`Initializing tsconfig for ${entryPoint.moduleId}`);
+    const basePath = path.dirname(entryPoint.entryFilePath);
 
-  // Resolve defaults from DI token and create a deep copy of the defaults
-  let tsConfig: TsConfig = JSON.parse(JSON.stringify(defaultTsConfig));
+    // Resolve defaults from DI token and create a deep copy of the defaults
+    let tsConfig: TsConfig = JSON.parse(JSON.stringify(defaultTsConfig));
 
-  // minimal compilerOptions needed in order to avoid errors, with their associated default values
-  // some are not overrided in order to keep the default associated TS errors if the user choose to set incorrect values
-  const requiredOptions: Partial<ts.CompilerOptions> = {
-    emitDecoratorMetadata: true,
-    experimentalDecorators: true,
-    moduleResolution: ts.ModuleResolutionKind.NodeJs,
-    target: ts.ScriptTarget.ES2015,
-    declaration: true,
-    lib: ['dom', 'es2015']
-  };
+    // minimal compilerOptions needed in order to avoid errors, with their associated default values
+    // some are not overrided in order to keep the default associated TS errors if the user choose to set incorrect values
+    const requiredOptions: Partial<ts.CompilerOptions> = {
+      emitDecoratorMetadata: true,
+      experimentalDecorators: true,
+      moduleResolution: ts.ModuleResolutionKind.NodeJs,
+      target: ts.ScriptTarget.ES2015,
+      declaration: true,
+      lib: ['dom', 'es2015']
+    };
 
-  const overrideConfig: Partial<TsConfig> = {
-    rootNames: [entryPoint.entryFilePath],
-    options: {
-      flatModuleId: entryPoint.moduleId,
-      flatModuleOutFile: `${entryPoint.flatModuleFile}.js`,
-      basePath: basePath,
-      rootDir: basePath,
-      outDir: '',
-      lib: entryPoint.languageLevel ? entryPoint.languageLevel.map(lib => `lib.${lib}.d.ts`) : tsConfig.options.lib,
-      // setting this as basedir will rewire triple-slash references
-      declarationDir: basePath,
-      // required in order to avoid "ENOENT: no such file or directory, .../.ng_pkg_build/..." errors when using the programmatic API
-      inlineSources: true,
-      // setting the below here because these are a must have with these valus
-      inlineSourceMap: true,
-      sourceMap: false,
-      sourceRoot: `ng://${entryPoint.moduleId}`
+    const overrideConfig: Partial<TsConfig> = {
+      rootNames: [entryPoint.entryFilePath],
+      options: {
+        flatModuleId: entryPoint.moduleId,
+        flatModuleOutFile: `${entryPoint.flatModuleFile}.js`,
+        basePath: basePath,
+        rootDir: basePath,
+        outDir: '',
+        lib: entryPoint.languageLevel ? entryPoint.languageLevel.map(lib => `lib.${lib}.d.ts`) : tsConfig.options.lib,
+        // setting this as basedir will rewire triple-slash references
+        declarationDir: basePath,
+        // required in order to avoid "ENOENT: no such file or directory, .../.ng_pkg_build/..." errors when using the programmatic API
+        inlineSources: true,
+        // setting the below here because these are a must have with these valus
+        inlineSourceMap: true,
+        sourceMap: false,
+        sourceRoot: `ng://${entryPoint.moduleId}`
+      }
+    };
+
+    tsConfig.rootNames = overrideConfig.rootNames;
+    tsConfig.options = { ...requiredOptions, ...tsConfig.options, ...overrideConfig.options };
+
+    switch (entryPoint.jsxConfig) {
+      case 'preserve':
+        tsConfig.options.jsx = ts.JsxEmit.Preserve;
+        break;
+      case 'react':
+        tsConfig.options.jsx = ts.JsxEmit.React;
+        break;
+      case 'react-native':
+        tsConfig.options.jsx = ts.JsxEmit.ReactNative;
+        break;
+      default:
+        break;
     }
-  };
 
-  tsConfig.rootNames = overrideConfig.rootNames;
-  tsConfig.options = { ...requiredOptions, ...tsConfig.options, ...overrideConfig.options };
+    // Add paths mappings for dependencies
+    const entryPointDeps = entryPoints.filter(x => x.data.entryPoint.moduleId !== entryPoint.moduleId);
+    if (entryPointDeps.length > 0) {
+      if (!tsConfig.options.paths) {
+        tsConfig.options.paths = {};
+      }
 
-  switch (entryPoint.jsxConfig) {
-    case 'preserve':
-      tsConfig.options.jsx = ts.JsxEmit.Preserve;
-      break;
-    case 'react':
-      tsConfig.options.jsx = ts.JsxEmit.React;
-      break;
-    case 'react-native':
-      tsConfig.options.jsx = ts.JsxEmit.ReactNative;
-      break;
-    default:
-      break;
-  }
+      for (let dep of entryPointDeps) {
+        const { entryPoint, destinationFiles } = dep.data;
+        const { moduleId, entryFilePath } = entryPoint;
+        const mappedPath = [destinationFiles.declarations, entryFilePath];
 
-  return tsConfig;
+        if (!tsConfig.options.paths[moduleId]) {
+          tsConfig.options.paths[moduleId] = mappedPath;
+        } else {
+          tsConfig.options.paths[moduleId].concat(mappedPath);
+        }
+      }
+    }
+
+    currentEntryPoint.data.tsConfig = tsConfig;
+  });
 };

